include/pybind11/async.h

```diff
@@ -0,0 +1,208 @@
+#pragma once
+
+#include <memory>
+#include <future>
+#include <chrono>
+
+#include "pybind11/pybind11.h"
+
+
+namespace py = pybind11;
+
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+PYBIND11_NAMESPACE_BEGIN(async)
+
+
+class StopIteration : public py::stop_iteration {
+    public:
+        StopIteration(py::object result) : stop_iteration("--"), result(result) {};
+        //using py::stop_iteration::stop_iteration;
+        /// Set the error using the Python C API
+        void set_result(py::object result) {
+            this->result = result;
+        }
+        void set_error() const override {
+            PyErr_SetObject(PyExc_StopIteration, this->result.ptr());
+        }
+    private:
+        py::object result;
+};
+
+
+class Awaitable : public std::enable_shared_from_this<Awaitable>{
+    public:
+        Awaitable() {
+            this->future = std::future<py::object>();
+        };
+
+        Awaitable(std::future<py::object>& _future) {
+            this->future = std::move(_future);
+        };
+
+        std::shared_ptr<Awaitable> iter() {
+            return this->shared_from_this();
```

Comment:
pybind11 maintains a registry of instances. I think you could simply `return this;`, it will reuse the existing Python instance. Then you also don't need to inherit from enable_shared_from_this and you can just use the default holder.

Reply:
Thanks for the hint! i updated the pr accordingly